File: minWindowSubstring.java
import java.util.*;

public class MinWindowSubstring {
    public static class Solution {
        public String minWindow(String s, String t) {
            if (s == null || t == null || s.length() < t.length()) return "";
            int[] freq = new int[128];
            for (char c : t.toCharArray()) freq[c]++;

            int l = 0, r = 0, minLen = Integer.MAX_VALUE, cnt = 0;
            int startInd = -1;
            int needed = t.length();

            while (r < s.length()) {
                char rc = s.charAt(r);
                if (freq[rc] > 0) cnt++;
                freq[rc]--;

                while (cnt == needed) {
                    if (r - l + 1 < minLen) {
                        minLen = r - l + 1;
                        startInd = l;
                    }
                    char lc = s.charAt(l);
                    freq[lc]++;
                    if (freq[lc] > 0) cnt--;
                    l++;
                }
                r++;
            }

            return startInd == -1 ? "" : s.substring(startInd, startInd + minLen);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.hasNextLine() ? sc.nextLine() : "";
        String t = sc.hasNextLine() ? sc.nextLine() : "";
        Solution sol = new Solution();
        System.out.println(sol.minWindow(s, t));
        sc.close();
    }
}

File: minWindowSubstring.cpp
#include <bits/stdc++.h>
using namespace std;

string minWindow(const string &s, const string &t) {
    int m = t.size();
    unordered_map<char,int> freq;
    for (char c : t) freq[c]++;

    int l = 0, r = 0, minLen = INT_MAX, cnt = 0;
    int startInd = -1;
    while (r < (int)s.size()) {
        if (freq[s[r]] > 0) cnt++;
        freq[s[r]]--;

        while (cnt == m) {
            if (r - l + 1 < minLen) {
                minLen = r - l + 1;
                startInd = l;
            }
            freq[s[l]]++;
            if (freq[s[l]] > 0) cnt--;
            l++;
        }
        r++;
    }
    return startInd == -1 ? string("") : s.substr(startInd, minLen);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string s, t;
    if (!getline(cin, s)) return 0;
    if (!getline(cin, t)) return 0;

    cout << minWindow(s, t) << '\n';
    return 0;
}
